pkg/capabilities/consensus/ocr3/median/median_aggregator.go

```diff
@@ -0,0 +1,256 @@
+		if currStaleness > int64(a.config.Heartbeat) ||
+			currDeviation > a.config.Deviation.InexactFloat64() {
+			previousReportInfo.ObservationTimestamp = currTimestamp
+			previousReportInfo.BenchmarkPrice = newPrice.Bytes()
+
+			feedIDBytes := a.config.FeedID.Bytes()
+			remappedID := a.config.RemappedID
+			if len(remappedID) == 0 { // if not provided, fall back to original ID
+				remappedID = feedIDBytes[:]
+			}
+			toWrap = append(toWrap,
+				map[string]any{
+					FeedIDOutputFieldName:     feedIDBytes[:],
+					RawReportOutputFieldName:  nil,
+					PriceOutputFieldName:      newPrice,
+					TimestampOutputFieldName:  currTimestamp,
+					RemappedIDOutputFieldName: remappedID,
+				})
+
+			shouldReport = true
+		}
+	}
+
+	marshalledState, err := proto.MarshalOptions{Deterministic: true}.Marshal(currentState)
+	if err != nil {
+		return nil, err
+	}
+
+	wrappedReportsNeedingUpdates, err := values.NewMap(map[string]any{
+		TopLevelListOutputFieldName: toWrap,
+	})
+	if err != nil {
+		return nil, err
+	}
+	reportsProto := values.Proto(wrappedReportsNeedingUpdates)
+
+	a.lggr.Debugw("Aggregate complete", "shouldReport", shouldReport)
+	return &types.AggregationOutcome{
+		EncodableOutcome: reportsProto.GetMapValue(),
+		Metadata:         marshalledState,
+		ShouldReport:     shouldReport,
+	}, nil
+}
+
+func (a *medianAggregator) initializeCurrentState(previousOutcome *types.AggregationOutcome) (*MedianOutcomeMetadata, error) {
+	currentState := &MedianOutcomeMetadata{}
+	if previousOutcome != nil {
+		err := proto.Unmarshal(previousOutcome.Metadata, currentState)
+		if err != nil {
+			return nil, err
+		}
+	}
+	// initialize empty state for missing feeds
+	if currentState.FeedInfo == nil {
+		currentState.FeedInfo = make(map[string]*MedianReportInfo)
+	}
+	feedID := a.config.FeedID
+	if _, ok := currentState.FeedInfo[feedID.String()]; !ok {
+		currentState.FeedInfo[feedID.String()] = &MedianReportInfo{
+			ObservationTimestamp: 0, // will always trigger an update
+			BenchmarkPrice:       big.NewInt(0).Bytes(),
+		}
+		a.lggr.Debugw("initializing empty onchain state for feed", "feedID", feedID.String())
+	}
+	a.lggr.Debugw("current state initialized", "state", currentState, "previousOutcome", previousOutcome)
+	return currentState, nil
+}
+
+func (a *medianAggregator) extractPayloads(observations map[ocrcommon.OracleID][]values.Value, aggregationKey string) [][]byte {
+	var events [][]byte
+	for nodeID, nodeObservations := range observations {
+		// TODO: check the following comment:
+		// we only expect a single observation per node
+		if len(nodeObservations) == 0 || nodeObservations[0] == nil {
+			a.lggr.Warnf("node %d contributed with empty observations", nodeID)
+			continue
+		}
+		if len(nodeObservations) > 1 {
+			a.lggr.Warnf("node %d contributed with more than one observation", nodeID)
+			continue
+		}
+		payload := &capabilities.CapabilityResponse{}
```

Comment:
This isn't correct, the observation would just be a values.Value but we wouldn't know exactly what it is, so we'll need to do some iterating over the types (or alternatively we can assume it's a map if users provide an aggregationKey).